### DISCUSSION TITLE
Mitigating untrusted file upload security issues

### DISCUSSION THREAD

**Discussion by RealOrangeOne** [wagtail/wagtail#12617] [Development]
Through the Wagtail Images and Documents features, it's possible for Admin users to upload arbitrary files, and have them served through the website. Serving files to end users has a number of security considerations, as the files may contain unwanted or unexpected content.

For Images, Wagtail has a few protections to ensure  that only valid images may be uploaded and served. However, the validation for documents is much simpler. Wagtail's Documents feature is intentionally designed to allow uploading and serving arbitrary files. Unfortunately, this flexibility comes with risks:

- Serving HTML, SVG or PDFs with embedded Javascript (Cross-site scripting, or XSS)
- Malformed images causing browser crashes (since image validation isn't performed)
- Serving HTML to spoof other pages of the site

(This list should not be considered exhaustive)

---

Documents can be served by Wagtail in a few ways (as configured by `WAGTAILDOCS_SERVE_METHOD`).  Some are served directly by Wagtail, whilst others are deferred to external storage (such as AWS S3). 

It's currently Wagtail's stance that it's up to the user to protect against these risks, although a few methods are [documented](https://docs.wagtail.org/en/stable/deployment/under_the_hood.html#documents):

- Serve all files as "attachments" (using the `Content-Disposition` header), rather than displaying in the browser. This forces the file to be downloaded instead, at the cost of user experience when opening PDFs (which many assume will open in the browser without downloading).
- Serve all files with a restricted `Content-Security-Policy`, which prevents inline resources from executing (CSS, JS etc).

Neither of these approaches can be implemented when serving files directly from remote storage ("direct" or "redirect") since there is no unified API for controlling the headers are sent when remote storage serves a file. It's likely infeasible for Wagtail to maintain integrations with every storage provider to facilitate this.

---

Wagtail has received multiple security vulnerability reports for this issue in the past (thank you to Ali İltizar and Georgios Roumeliotis). Our response has been that it's down to the user to secure. Whilst document uploads are generally restricted to Wagtail Admin users, who are likely trusted, Wagtail should help the user wherever possible to ensure their site is secure.

This discussion is opened to try and collect potential solutions to this problem.

**Comment by thibaudcolas** [comment]
I thought I’d directly share the suggested mitigations from the recent report of security researcher @alii76tt (thank you!) as they’re all very excellent:

## 1. Validate File Contents

> In the models.py file, the open_file function should include logic to check for and sanitize potentially malicious content:

```python
@contextmanager
def open_file(self):
    ...
    try:
        # Scan file content for malicious scripts
        file_content = f.read()
        if b'/JavaScript' in file_content or b'app.alert' in file_content:
            raise ValueError("Malicious content detected in uploaded file.")
        yield f
    finally:
```

## 2. Implement File Type Restrictions

> Use libraries to validate the actual contents of the uploaded files (e.g., PyPDF2 for PDFs):

```python
from PyPDF2 import PdfReader

def validate_pdf(file):
    try:
        reader = PdfReader(file)
        for page in reader.pages:
            if '/JS' in page.get('/Annots', ''):
                raise ValueError("PDF contains embedded JavaScript.")
    except Exception:
        raise ValueError("Invalid or potentially malicious PDF.")
```

## 3. Disallow JavaScript in PDF Annotations

> Reject any file containing /JavaScript or /AA (additional actions).

## 4. Use Content Security Policies (CSP)

> Implement strong CSP headers to mitigate XSS:
> `Content-Security-Policy: script-src 'self'; object-src 'none';`

## 5. Notify Users and Log Incidents

> Log attempts to upload malicious files for auditing.
> Notify users when an upload fails due to content validation.

**Reply by thibaudcolas** [reply]
To me the ones that would be the most promising at this point in time are:

1. Support strict CSPs (lots of work already going towards that, #1288)
2. Have audit logs for file uploads within the CMS (see [media versioning #58](https://github.com/wagtail/roadmap/issues/58) on our roadmap)
3. Support integrations with file scanning / antivirus software. I believe at Torchbox we use [bucket/av](https://bucketav.com/).

**Comment by iltosec** [comment]
Thank you for including my suggestions in this discussion. It’s great to see Wagtail taking steps to enhance security. If you consider implementing one or more of these suggestions in the future, I’d be happy to assist with testing processes and implementation details.

**Comment by tomusher** [comment]
I'm not entirely clear on the potential attack vectors we're trying to mitigate here, perhaps you could share more from the report?

From what I can see, there are a number of vectors but the mitigations don't have much crossover:

1. [Content Spoofing](https://owasp.org/www-community/attacks/Content_Spoofing) with the intention of reputation hijacking - Malicious user can upload content to be served on a trusted domain, e.g. https://www.example.com/documents/phishing-form
2. [Cross-site Content Hijacking](https://github.com/nccgroup/CrossSiteContentHijacking) (XSCH) through PDFs (and other non-HTML file types) - Malicious user uploads file that can access session cookies/site data through vulnerabilities (or features) in a specific file format. Users accessing this resource directly leak session data.
3. [Cross-site Content Hijacking](https://github.com/nccgroup/CrossSiteContentHijacking) through HTML/JS - Malicious user uploads file that can access session cookies/site data through direct access to cookies in JS. Users accessing this resource directly leak session data.
4. The application of vectors 2 and 3 on content which is embedded in a Wagtail site page - seems highly dependent on how the site is built, generally seems unlikely that sites will be embeddeding content from documents directly.
5. Other vulnerabilities from [Unrestricted File Upload](https://owasp.org/www-community/vulnerabilities/Unrestricted_File_Upload) - e.g. uploading viruses, uploading large files, malfomed files, denial of service etc.

(please do share anything I've missed here)

Potential mitigations (some as discussed, some new):

1. Recommending documents be served with the `X-Content-Type-Options: nosniff` headers and the correct `Content-Type`. Prevents browsers from guessing the mimetype of a file, e.g. if someone uploads a PDF with a JPG extension, the browser won't override the Content-Type and will not try to serve it as a PDF. Not controllable by Wagtail, can only be documented.
2. Stricter control of uploaded file types - `WAGTAILDOCS_EXTENSIONS` uses quite a naive approach to restricting file uploads, relying on file extensions. Guessing mimetypes with something like [filetype](https://pypi.org/project/filetype/) would make this more resilient.
3. Block uploading malicious files with virus scanner - would catch some malicious files but likely not XSS.
4. Block uploading malicious files with additional validation heuristics - detecting common malicious patterns in uploaded files. Downside here is that this is a constantly moving target and likely beyond the scope of what Wagtail should be doing.
5. Add CSP headers to Wagtail served content - can be controlled by Wagtail
7. Add CSP headers to served documents - e.g. served via S3/CDN, as mentioned, can only be documented, not controllable by Wagtail.
 
|                                  | 1. nosniff | 2. upload control | 3. virus scanner | 4. validation heuristics | 5. Wagtail CSP | 6. Document CSP |
|----------------------------------|------------|-------------------|------------------|--------------------------|----------------|-----------------|
| 1. Content Spoofing (reputation) |            | ✅ (partial)       |                  |                          |                |                 |
| 2. XSCH (PDF)                    | ✅          | ✅ (partial)       | ✅ (partial)      | ✅ (partial)              |                |                 |
| 3. XSCH (HTML)                   | ✅          | ✅ (partial)       |                  | ✅ (partial)              |                | ✅               |
| 4. Embedded Content              | ✅          | ✅ (partial)       |                  |                          | ✅ (partial)    |                 |
| 5. Malicious files               |            | ✅ (partial)       | ✅                | ✅ (partial)              |                |                 |

One approach that should cover most vectors would be to:

- Improve file detection at upload time, restricting uploaded files to a list of given mimetypes rather than extensions.
- Add a default restriction on files that can be uploaded. There is currently no default value for `WAGTAILDOCS_EXTENSIONS` (only a default value set in the project template). The ability to upload e.g. javascript, HTML and executable files seems like it should always require an explicit opt-in.
- Provide hooks for file validation to allow developers to insert their own checks (virus scanning, file content heuristics)
- Document `nosniff` and CSP recommendations, warn against allowing e.g. HTML file uploads.

I'm all for the audit logging plan too!

**Reply by gasman** [reply]
> Recommending documents be served with the `X-Content-Type-Options: nosniff` headers and the correct `Content-Type`. Prevents browsers from guessing the mimetype of a file, e.g. if someone uploads a PDF with a JPG extension, the browser won't override the Content-Type and will not try to serve it as a PDF. Not controllable by Wagtail, can only be documented.

To some extent we _can_ control this - in the out-of-the-box configuration, documents are served through a Django view, primarily to enforce authorization checks, but this also gives us the opportunity to set our own HTTP headers. In particular, most documents (by default, all except PDF) are served with `Content-Disposition: attachment`, so a browser will download them rather than open them and execute scripts. However, this is out of our control if the site owner omits to lock down direct access to the `/media/documents/` folder and the attacker links to the file there, or if the site owner configures Wagtail to skip the Django view layer (although usually you'd only do this if you were hosting documents somewhere remote like S3, and at that point any malicious scripts in them have no privileges over the Wagtail server). So, there are a lot of variables at play...

> Add a default restriction on files that can be uploaded. There is currently no default value for `WAGTAILDOCS_EXTENSIONS` (only a default value set in the project template). The ability to upload e.g. javascript, HTML and executable files seems like it should always require an explicit opt-in.

We spent some time weighing this up during #11735. Since there's no exhaustive list of potentially-exploitable file types, any opt-in mechanism aimed at disallowing JS/HTML/exe by default would have to equally apply to the multitude of non-scriptable file types that a user might reasonably upload as a document, but aren't common enough to be on any "allow list" that Wagtail might provide out-of-the-box.

A key point to bear in mind is that for a large portion of Wagtail's user base, the possibility of a malicious editor on staff is non-existent within their risk profile (for example, if it's a one-person shop where the owner of the only account is also the server administrator). For these users, it's entirely legitimate for them to configure Wagtail so that they can upload any new file type as the need arises, without having to individually allow it in the site config - from their perspective, the whole point of having a Wagtail installation is to let them publish whatever they want, without it trying to protect them from themselves...

The `WAGTAILDOCS_EXTENSIONS` line in the project template, and accompanying code comment, serves as a sort of checkbox to say "by deleting this line I declare that I don't have to worry about malicious editors", and I do think that's an appropriate level of risk mitigation.

**Comment by thibaudcolas** [comment]
> vulnerabilities (or features) in a specific file format

🥲 made me chuckle. Thank you for the excellent analysis @tomusher! Re audit logging, I noted we actually already have that on our roadmap too: [Media versioning #58](https://github.com/wagtail/roadmap/issues/58)

---

> A key point to bear in mind is that for a large portion of Wagtail's user base, the possibility of a malicious editor on staff is non-existent within their risk profile

@gasman I wonder if this discussion might be a good one for us to run past our corporate / large institutional users, to get their thoughts on risk profile? I’m also pretty sure this isn’t high up on everyone’s list (there’s loads more nefarious things people could do), but possibly there are a few here and there who will have considered this and can tell us what’s worth documenting. I recall in particular a large bank adopting Wagtail and deciding they had to adopt audit logging for __all__ CMS-managed data in order to make Wagtail viable.

I’m also not sure in what context exactly Torchbox uses bucket/av. If this is ever for Wagtail-related content, could be worthwhile to at least document the existence of this kind of tool? (and packages would be neat, if there’s any reason to integrate that with the CMS).

**Comment by gasman** [comment]
If I'm not mistaken, it's not actually possible to execute an XSS attack against a Wagtail site from a PDF document hosted on it, even if the ability to pop up an `alert()` suggests that it might be. For an XSS attack to be successful, the JS runtime needs to:

1. have access to the victim's login session
2. be able to make network requests, either to perform actions with the victim's credentials, or to exfiltrate those credentials to a host controlled by the attacker

Clearly if the user _downloads_ the PDF and opens it in a standalone application such as Acrobat Reader, point 1 no longer applies, so we only need to be concerned with the embedded PDF viewers in browsers. Various sources show that both Chrome and Mozilla enforce restrictions on one or both of these points as a matter of policy:

* https://github.com/mozilla/pdf.js/issues/13284 ("no networking support is implemented")
* https://issues.chromium.org/issues/40082516 ("stubs out the dangerous objects and methods", and notes that an alert does not demonstrate an XSS attack and that a valid vulnerability report would have to "show access to cookies or other private data")
* https://www.mozilla.org/en-US/security/advisories/mfsa2024-21/#CVE-2024-4367 (arbitrary JS execution is handled as a high-severity vulnerability, which clearly establishes that it's not allowed by design)

Since browser-level vulnerabilities can exist, I would be in favour of adding a `Content-Security-Policy` header to Wagtail's document `serve` view as an additional safeguard. However, since there is no indication that an attack is possible _on an appropriately patched browser_ (and, of course, any vulnerabilities at the browser level are outside of Wagtail's responsibility), the lack of such a safeguard should not be considered as a security issue on Wagtail's part - and there are no grounds for making changes that would hurt the user or developer experience, such as serving PDFs as attachments or adding heuristic checks on PDF content (which may result in false positives - what if it's a PDF document _about_ malicious Javascript in PDFs? :-) )

XSS attacks in file types _other_ than PDF were previously addressed in https://github.com/wagtail/wagtail/pull/11735. The default `WAGTAILDOCS_EXTENSIONS` list for new projects excludes any file types besides PDF that allow embedded JS, and the accompanying code comment (and [documentation for the `WAGTAILDOCS_EXTENSIONS` setting](https://docs.wagtail.org/en/stable/reference/settings.html#wagtaildocs-extensions)) highlight the extra steps that the developer should take to secure the webserver if they change this. Since the `serve` view serves non-PDF documents with `Content-Disposition: attachment` as standard, a vulnerability only arises if the developer allows document serving to bypass this view - at which point it's outside the scope of any protective measures Wagtail could take. This may happen either accidentally by failing to block direct access to `/media/documents`, or intentionally by changing `WAGTAILDOCS_SERVE_METHOD` - and then only if all of the following are true:

* the developer has changed `WAGTAILDOCS_EXTENSIONS` to include a file type that allows arbitrary JS, such as HTML or SVG;
* the developer allows the `serve` view to be bypassed and has not implemented the suggested measures such as CSP or `Content-Disposition` headers;
* documents are being served from the same hostname as the Wagtail admin;
* Untrusted users have been given document upload permission within the Wagtail admin.

**Reply by gasman** [reply]
> what if it's a PDF document _about_ malicious Javascript in PDFs? :-)

Now I'm sad to discover that [contextis.com](https://web.archive.org/web/20161203012245/https://www.contextis.com//), the first Wagtail site I ever encountered in the wild outside of my work on Wagtail itself, no longer exists as they were bought out by Accenture in 2020 - since they're an organisation that could plausibly have published such a document :-)

**Reply by thibaudcolas** [reply]
The rationale makes sense to me!

> a valid vulnerability report would have to "show access to cookies or other private data"

Yes, seems like an important threshold to meet.

**Comment by thibaudcolas** [comment]
This was raised again in the [Independent security audit #14173](https://github.com/wagtail/wagtail/discussions/14173) we received. Direct link to [file upload sub-section](https://github.com/wagtail/wagtail/discussions/14173#user-content-finding-lack-of-server-side-checks-received-files), with three specific aspects that I’m not sure we have entirely figured out yet:

- Problems with filenames (like path traversal). I’m not aware of it being an issue on a properly-configured project in our stack 🤔 not sure how much we might want to accommodate misconfigured projects / hosting environments with some vulnerable components on the same server. I’m sure there are also Django-level mitigations to this that I’m not aware of, and of course mitigations at the storage backend level.
- Antivirus scanning. I’ve added a short section about this in our docs in #14179, just documenting what is already feasible.
- File upload size. Lots of options here, and we are shipping `WAGTAILDOCS_MAX_UPLOAD_SIZE` in Wagtail 7.4. PR #14179 adds one section about it in the same document storage/serving docs, and I propose we set it to 10MB in our project template.

**Comment by thibaudcolas** [comment]
Noting from the report we received today, the discussion here also applies to SVGs uploaded as images and accessed directly. I have created a small docs tweak PR to make it clearer what protections we do offer: #14351.

I spent some time reconsidering if it might be worth sanitizing SVGs in _that_ upload pipeline but I just don’t see it. To provide good security guarantees when users open SVGs directly, we’d have to strip a lot of SVG features that are generally valuable when SVGs are loaded in `<img>` tags. The tradeoff just doesn’t seem worthwhile.
